helix-core/src/test/java/org/apache/helix/controller/stages/TestStateTransitionThrottleController.java

```diff
@@ -0,0 +1,193 @@
+        new StateTransitionThrottleConfig(ANY, StateTransitionThrottleConfig.ThrottleScope.INSTANCE,
+            anyInstanceLevelQuota),
+        new StateTransitionThrottleConfig(RECOVERY_BALANCE,
+            StateTransitionThrottleConfig.ThrottleScope.INSTANCE, recoveryInstanceLevelQuota),
+        new StateTransitionThrottleConfig(LOAD_BALANCE,
+            StateTransitionThrottleConfig.ThrottleScope.INSTANCE, loadInstanceLevelQuota),
+        new StateTransitionThrottleConfig(ANY, StateTransitionThrottleConfig.ThrottleScope.RESOURCE,
+            anyResourceLevelQuota),
+        new StateTransitionThrottleConfig(RECOVERY_BALANCE,
+            StateTransitionThrottleConfig.ThrottleScope.RESOURCE, recoveryResourceLevelQuota),
+        new StateTransitionThrottleConfig(LOAD_BALANCE,
+            StateTransitionThrottleConfig.ThrottleScope.RESOURCE, loadResourceLevelQuota));
+    ClusterConfig clusterConfig = new ClusterConfig("config");
+    clusterConfig.setStateTransitionThrottleConfigs(stateTransitionThrottleConfigs);
+
+    StateTransitionThrottleControllerAccessor controller =
+        new StateTransitionThrottleControllerAccessor(RESOURCE, INSTANCE, clusterConfig);
+
+    Assert.assertTrue(controller.isThrottleEnabled());
+
+    // verify behavior after charging cluster
+    controller.chargeCluster(ANY);
+    Assert.assertEquals(controller.getClusterLevelQuota(ANY), anyClusterLevelQuota - 1);
+    controller.chargeCluster(RECOVERY_BALANCE);
+    Assert.assertEquals(controller.getClusterLevelQuota(RECOVERY_BALANCE),
+        recoveryClusterLevelQuota - 1);
+    controller.chargeCluster(LOAD_BALANCE);
+    Assert.assertEquals(controller.getClusterLevelQuota(LOAD_BALANCE), loadClusterLevelQuota - 1);
+
+    // verify behavior after charging instance
+    controller.chargeInstance(ANY, INSTANCE);
+    Assert.assertEquals(controller.getInstanceLevelQuota(ANY, INSTANCE), anyInstanceLevelQuota - 1);
+    controller.chargeInstance(RECOVERY_BALANCE, INSTANCE);
+    Assert.assertEquals(controller.getInstanceLevelQuota(RECOVERY_BALANCE, INSTANCE),
+        recoveryInstanceLevelQuota - 1);
+    controller.chargeInstance(LOAD_BALANCE, INSTANCE);
+    Assert.assertEquals(controller.getInstanceLevelQuota(LOAD_BALANCE, INSTANCE),
+        loadInstanceLevelQuota - 1);
+
+    // verify behavior after charging resource
+    controller.chargeResource(ANY, RESOURCE);
+    Assert.assertEquals(controller.getResourceLevelQuota(ANY, RESOURCE), anyResourceLevelQuota - 1);
+    controller.chargeResource(RECOVERY_BALANCE, RESOURCE);
+    Assert.assertEquals(controller.getResourceLevelQuota(RECOVERY_BALANCE, RESOURCE),
+        recoveryResourceLevelQuota - 1);
+    controller.chargeResource(LOAD_BALANCE, RESOURCE);
+    Assert.assertEquals(controller.getResourceLevelQuota(LOAD_BALANCE, RESOURCE),
+        loadResourceLevelQuota - 1);
+  }
+
+  private static class StateTransitionThrottleControllerAccessor
```

Comment:
The convention is to prepend "Mock" -> MockStateTransitionThrottleControllerAccessor in Helix's codebase.

But I'm not too particular about the naming of the class. But I think you should add a block comment here why you have this class. Just saying that you are extending it to expose some internals for testing is fine?

Reply:
Added the comment. "Mock" doesn't suit the use case here because it mainly refers to "mocking the class and mutate some behaviors at will"